DX: Progress output refactor

Wirone opened the pull request:
Extracted from #6842 for easier review. New progress outputs will be provided in that MR, will be rebased after merging this one.

This PR does not bring new features, only prepares the code to be extended with new progress formats.

Wirone commented on src/Console/Output/OutputContext.php:
```diff
@@ -0,0 +1,49 @@
+
+use Symfony\Component\Console\Output\OutputInterface;
+
+final class OutputContext
```
Missing `@internal`

Wirone commented on src/Console/Command/FixCommand.php:
```diff
@@ -278,22 +279,21 @@ protected function execute(InputInterface $input, OutputInterface $output): int
-            $progressOutput = new ProcessOutput(
+        $progressType = $resolver->getProgressType();
+        $progressOutput = ProgressOutputFactory::create(
+            $resolver->getProgressType(),
```
```suggestion
            $progressType,
```

Wirone reviewed (commented):
Self-review 😉

keradus commented:
[kudos for commit logs]

keradus commented on src/Console/Output/ProcessOutputInterface.php:
```diff
@@ -19,5 +19,12 @@
  */
 interface ProcessOutputInterface
 {
+    public const OUTPUT_TYPE_NONE = 'none';
```
I am up for having the constants/enums for those values.

Yet, I do not find it proper to have them declared in the interface.

`class FooOutput implements ProgressOutputInterface` - why would `FooOutput` be aware `OUTPUT_TYPE_DOTS` , at all? And then, when someone would add `FooOutput`, why they would need to touch the interface to add the `OUTPUT_TYPE_FOO` _to interface_ ?
and then imagine that it's non-internal class and one would like to register external , custom ProgressOutput class - and boom, they cannot adjust interface at source repo.

This should be outside of the interface and allow for extensions. Maybe mimic what we do for Fixers (and it's possibility to register custom fixers - not saying we need to introduce that possibility in this PR, but let's have a solution that will allow for that in future)

keradus commented on src/Console/Output/Progress/ProgressOutputInterface.php:
```diff
@@ -0,0 +1,34 @@
+ */
+interface ProgressOutputInterface
+{
+    public const OUTPUT_TYPE_NONE = 'none';
+    public const OUTPUT_TYPE_DOTS = 'dots';
+    public const OUTPUT_TYPES = [
+        self::OUTPUT_TYPE_NONE,
+        self::OUTPUT_TYPE_DOTS,
+    ];
```
I am up for having the constants/enums for those values.

Yet, I do not find it proper to have them declared in the interface.

`class FooOutput implements ProgressOutputInterface` - why would `FooOutput` be aware `OUTPUT_TYPE_DOTS` , at all? And then, when someone would add `FooOutput`, why they would need to touch the interface to add the `OUTPUT_TYPE_FOO` _to interface_ ?
and then imagine that it's non-internal class and one would like to register external , custom ProgressOutput class - and boom, they cannot adjust interface at source repo.

This should be outside of the interface and allow for extensions. Maybe mimic what we do for Fixers (and it's possibility to register custom fixers - not saying we need to introduce that possibility in this PR, but let's have a solution that will allow for that in future)

Wirone replied:
Agree, moved to `OutputProgressType` class.

keradus replied:
this by no means addresses the majority of my initial comment
> And then, when someone would add FooOutput, why they would need to touch the interface to add the OUTPUT_TYPE_FOO to ~interface~ *enum* ?
and then imagine that it's non-internal class and one would like to register external , custom ProgressOutput class - and boom, they cannot adjust interface at source repo.
>
> This should be outside of the interface and allow for extensions. Maybe mimic what we do for Fixers (and it's possibility to register custom fixers - not saying we need to introduce that possibility in this PR, but let's have a solution that will allow for that in future)

Wirone replied:
See [this comment](https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/pull/6848#discussion_r1186573954).

keradus commented on src/Console/Output/OutputContext.php:
what is the value of this DTO?

Wirone replied:
It carries information about output environment from `FixCommand` to output writers. Previously it was used only in `ProcessOutput` (now `DotsOutput`), but in #6842 new output writers are introduced (along with support for custom writers) and they also need this information. So it's wrapped in the DTO and passed through a factory, so writers can then use it to align output properly.

keradus commented on src/Console/Command/FixCommand.php:
```diff
@@ -302,9 +302,11 @@ protected function execute(InputInterface $input, OutputInterface $output): int
             $resolver->shouldStopOnViolation()
         );
 
+        $this->eventDispatcher->addListener(FixerFileProcessedEvent::NAME, [$progressOutput, 'onFixerFileProcessed']);
```
why subscribing to event handling was taken outside of ProgressOutput?

Wirone replied:
Because it does not make sense to inject event dispatcher to output writer. Previously dispatcher was injected there only for it to be invoked internally for attaching and detaching listener. It's not the writer's responsibility. Instead, `ProgressOutputInterface` defines a contract between the `FixCommand` and output writers.

Wirone replied:
Also, as I said above, other output writers are added in #6842 and they also need to listen on that event, it does not make sense to inject event dispatcher to all writers and copy&paste attach/detach logic.

keradus replied:
fair point 👍🏻 
that reason was _not_ visible in commit log

keradus replied:
hmm, what I wonder here @Wirone .

Originally, listener was registered only if it was needed - now, it's registered even if `ProgressOutputType::NONE`

keradus commented on src/Console/Output/ProcessOutputFactory.php:
```diff
@@ -0,0 +1,46 @@
+
+final class ProcessOutputFactory
+{
+    public static function create(string $outputType, OutputContext $context): ProcessOutputInterface
```
no static methods pls.

static is as evil as global.
good luck for whoever later to introduce proper DIC or mocking ProcessOutputFactory as dependency.

keradus commented on src/Console/Output/ProcessOutputFactory.php:
```diff
@@ -0,0 +1,46 @@
+
+final class ProcessOutputFactory
+{
+    public static function create(string $outputType, OutputContext $context): ProcessOutputInterface
```
no static methods pls.

static is as evil as global.
good luck for whoever later to introduce proper DIC or mocking ProcessOutputFactory as dependency.

Wirone replied:
Good luck with mocking current `FixCommand`'s dependencies too:

```php
public function __construct(ToolInfoInterface $toolInfo)
{
    parent::__construct();

    $this->eventDispatcher = new EventDispatcher();
    $this->errorsManager = new ErrorsManager();
    $this->stopwatch = new Stopwatch();
    $this->defaultConfig = new Config();
    $this->toolInfo = $toolInfo;
}
```

😉 

`ProgressOutputFactory` is used within `FixCommand` too, this is the same level of non-interoperability as current objects initialised in the constructor. What's the point in making it non-static and creating instance in the command's constructor, just to call one factory method? 🤷‍♂️ 

But yeah, if you really want, I can change it, even though I don't see much sense in it.

kubawerlos replied:
> no static methods pls.
> 
> static is as evil as global.

quite a subjective opinion, there is nothing wrong with reasonable static method usage: https://verraes.net/2014/06/when-to-use-static-methods-in-php/

keradus replied:
if Foo is ugly, it's not a reason that newly introduce Bar should be ugly too.

kubawerlos replied:
I'd advocate that stateless static function is much less ugly that using `new` in constructor.

keradus replied:
i decided to limit myself to ask for improvement in code introduced in this PR, not to ask to improve whole codebase ;)

Wirone replied:
OK, so what's the conclusion? Remove `static` from method and initialise `ProgressOutputFactory` in the `FixCommand`'s constructor just like other dependencies? I would like to minimise risk of another round of review 😉

keradus replied:
yep.
and then ultimately [in one other PR, in 2030], stop doing it in constructor but inject it from DIC.

keradus commented on tests/Console/Output/OutputContextTest.php:
```diff
@@ -0,0 +1,43 @@
+            $output = new NullOutput(),
+            $dispatcher = new EventDispatcher(),
+            $width = 100,
+            $filesCount = 10
```
I consider assigning values in conditions, in passing parameters etc to be a side effect.
please, don't rely on side effects.

keradus commented on tests/Console/Output/ProcessOutputFactoryTest.php:
```diff
@@ -0,0 +1,66 @@
+
+        yield 'none' => ['none', $context, NullOutput::class];
+
+        yield 'dots' => ['dots', $context, ProcessOutput::class];
```
why not using `OUTPUT_TYPE_DOTS` ?

Wirone replied:
You mean something like this?

```php
        yield 'none' => [ProgressOutputType::NONE, $context, NullOutput::class];

        yield 'dots' => [ProgressOutputType::DOTS, $context, ProcessOutput::class];
```

I wasn't sure if this should be dynamic thing or explicit value. Like in migrations, you shouldn't use dynamic content from the code, rather explicit values representing state at particular point in time, so they don't break when e.g. constant is changed. If actual values (constants) are changed then test fails and requires manual action which is rather good - otherwise it could still work but could make no sense, like:

```php
        yield 'dots' => [ProgressOutputType::SOME_RENAMED_CONST, $context, ProcessOutput::class];
```

Key does not correspond with progress type here. But it's not strong opinion, I don't have established standard for that, so only thinking out loud 😉.

keradus replied:
If someone would replace value under constant, I wouldn't like to need to update all tests to reflect new value. That's exactly why we have the constant - to not use value itself, again and again, and then manually update everywhere.

about the yielding key, I'm actually OK to have it follow the constant NAME rather than it's value.

Wirone replied:
> That's exactly why we have the constant - to not use value itself, again and again, and then manually update everywhere.

I agree, in terms of the production code. Not sure about tests 🤷‍♂️. But OK, I'll change it 👍

keradus commented:
@Wirone , please avoid destructing pushes (eg force-pushes like squashing, rebases) in PRs for this repo. it makes it harder for reviewer to check what they already reviewed and what not

keradus commented on tests/Console/Output/OutputContextTest.php:
```diff
@@ -15,9 +15,8 @@
 
 use PhpCsFixer\Console\Output\OutputContext;
-use PHPUnit\Framework\TestCase;
+use PhpCsFixer\Tests\TestCase;
```
*observation*

so nice commits history and then boom, this change, totally not related to commit log ;P

keradus commented on tests/Console/Output/DotsOutputTest.php:
```diff
@@ -15,24 +15,24 @@
+ * @covers \PhpCsFixer\Console\Output\DotsOutput
  */
-final class ProcessOutputTest extends TestCase
+final class DotsOutputTest extends TestCase
```
*idea* 
we could consider to create base test class for XyzOutputTests, to unify what and how do we test.
We used this approach here: https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/blob/master/tests/Console/Report/FixReport/AbstractReporterTestCase.php

Wirone replied:
That's [what I did](https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/pull/6842/files#diff-005c12ac66ec8494e12d238906f3e1057b2d78a94be8e2665efc29c6b139d305R22) in #6842 where more test-related code is shared between several output writers.

keradus replied:
would it make sense to have the tests refactoring addressed in refactoring PR and not in introducing new feature PR?

keradus commented on tests/Console/Output/Progress/DotsOutputTest.php:
```diff
@@ -179,7 +179,7 @@ public static function provideProcessProgressOutputCases(): array
     {
         $this->expectException(\BadMethodCallException::class);
-        $this->expectExceptionMessage('Cannot serialize PhpCsFixer\Console\Output\DotsOutput');
+        $this->expectExceptionMessage('Cannot serialize PhpCsFixer\Console\Output\Progress\DotsOutput');
```
use `DotsOutput::class`

Wirone replied:
FYI This is code from original `ProcessOutput`, not mine, so I just adjusted it to new FQCN but did not change the approach. But ok, I can change it.

keradus replied:
got you, happy with both then - agree it could be out of scope of PR. hard to notice when github shows as new file 😓 
[ultimately I believe we should use `::class`, like we already doing in majority of places, but it wasn't easy to find all cases to improve]

keradus commented on src/Console/Output/Progress/ProgressOutputFactory.php:
```diff
@@ -24,20 +27,25 @@ public static function create(string $outputType, OutputContext $context): Progr
-                return new DotsOutput($context);
+        return ProgressOutputInterface::OUTPUT_TYPE_NONE === $outputType
+            ? new NullOutput()
+            : new DotsOutput($context);
```
in general, `switch-case` is using `==` instead of `===` , thus forbidden [same as we forbid `==`. but also, imagine having 20 options inside. ternary or if-elseif-elseif-l.... will be insane.

```php
factoryMap = [
    null: ($context) => new NullOutput(),
    dots: ($context) => new DotsOutput($context),
    ...
];

return factoryMap[$outputType]($context); // O(1)
```

Wirone replied:
That's [what I do](https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/pull/6842/files#diff-166b79b3f269a4ecbc711d6f2ae4da497b7448c1bc5ba2d688cd8e94cda0e8e0R27-R32) in #6842 where more output types were added 🙂. Here simple ternary is enough.

keradus replied:
I would be OK to see that part of refactoring already in refactoring PR, as... purpose of refactoring PR should be refactoring for future needs - otherwise you refactor a bit here, and a bit there.

but non-blocking

keradus commented:
@Wirone , github shows comments of mine as outdated, but please look into them - you moved the file to directory and GH got lost

Wirone commented:
> github shows comments of mine as outdated, but please look into them - you moved the file to directory and GH got lost

@keradus most probably you were reviewing single commits and added comments to already outdated code, because the only thing I force-pushed was amend to "[Small fixes caught on self-review](https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/pull/6848/commits/b88ab20a18a7870a4245ff7cce4931da017b29fd)" where I changed `@interal` to `@internal` (found the typo when working on fixes based on your comments). Previously provided code was not modified and files weren't moved 🙂.

keradus commented:
eg https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/pull/6848#discussion_r1146894122

I commented on 
`tests/Console/Output/DotsOutputTest.php`

now file is 
`tests/Console/Output/Progress/DotsOutputTest.php`

so GH believes my comment is outdated

Wirone commented:
@keradus I know, but `tests/Console/Output/Progress/DotsOutputTest.php` was here from the very beginning of this PR. It was moved in "[Extract progress-related classes to separate namespace](https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/pull/6848/commits/d0f2bb16d56a456a2580d03697e5344559a203c8)" which was pushed along with other commits (as you can see at the top ot the timeline, CI tick is next to subsequent commit).

For example [this comment](https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/pull/6848#pullrequestreview-1355672402) links to [this revision](https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/pull/6848/files/cb0071d5e29c032f9eadf21b84dab04a6678ba8f..5b885c143a343f041b00ada2441b6948c10ee077#diff-c5683353a5d6bf590a34d7811b468561d23fd5258a30d4f70eee3d45e30e3458) (when you click commented file's name) where you clearly can see it's diff for one commit, which's content was already outdated at the time of writing comment 😉.

<img width="1196" alt="image" src="https://user-images.githubusercontent.com/600668/227383548-511ce909-d0db-4989-aeb3-93deba944f51.png">

Commit log is good for reviewing thought process, but for reviewing actual code only "Files changed" tab should be used, with newest diff.

keradus commented on src/Console/Output/Progress/ProgressOutputType.php:
```diff
@@ -0,0 +1,19 @@
+/**
+ * @internal
+ */
+final class ProgressOutputType
```
*major*

to allow register custom ProgressOutput implementation, we should get rid of this enum class.
context: https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/pull/6848#discussion_r1146888035

I'm OK to merge this PR without this change implemented, to move on forward. But we should not keep it in PR that will allow for custom implementation

Wirone replied:
Changed my mind, I will keep it here and let's continue in #6842. Since it's `@internal` it can be safely changed/removed later, if we decide like that.

But in short: this class only describes built-in output types, so there's no need for custom output classes to be registered here, they are going to be handled [differently](https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/pull/6842#discussion_r1160323860)

keradus replied:
overall, I do believe that we should have similar mechanism to https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/blob/master/src/Console/Report/FixReport/ReporterInterface.php#L24 , rather than centralised enum-like class aiming to collect all possible types

keradus commented:
tbh I failed to pick up your point.

My point: github marks this comment https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/pull/6848#discussion_r1146880761 as outdated. It is not. Please apply requested change.

keradus commented:
> Commit log is good for reviewing thought process, but for reviewing actual code only "Files changed" tab should be used, with newest diff.

I don't agree with `should` part myself.
you ask to review code twice, i prefer to avoid this. the bigger the diff, the more you need context of given change, instead of reading whole at once.

Wirone commented:
@keradus can we finally finish this one? 🙏

keradus commented on src/Console/Output/Progress/DotsOutput.php:
```diff
@@ -12,18 +12,18 @@
 
 /**
- * Output writer to show the process of a FixCommand.
+ * Output writer to show the process of a FixCommand using dots and meaningful letters.
```
process -> progress?

keradus commented on src/Console/Command/FixCommand.php:
```diff
@@ -297,9 +300,11 @@ protected function execute(InputInterface $input, OutputInterface $output): int
             $resolver->shouldStopOnViolation()
         );
 
+        $this->eventDispatcher->addListener(FixerFileProcessedEvent::NAME, [$progressOutput, 'onFixerFileProcessed']);
```
hm, I commented it out and no test failed

Wirone replied:
It was previously inside `ProcessOutput`, I just moved it here to bind events with any progress output (the one that is configured and used). I believe there wasn't any integration test for that, there are only unit tests for progress outputs that don't require event dispatcher, the `onFixerFileProcessed()` method is called directly.

keradus replied:
ref https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/pull/6848#issuecomment-1616171811

i do not understand what decision you look for here. what is the question?

Wirone replied:
If I should add such tests (IMHO out of the scope of this PR) or not 😉.

keradus replied:
i'm ok to have it non-blocking, can be out of scope, as already missing on master.

keradus commented on src/Console/Output/Progress/DotsOutput.php:
```diff
@@ -39,11 +39,7 @@ final class ProcessOutput implements ProcessOutputInterface
-    private EventDispatcherInterface $eventDispatcher;
-
-    private int $files;
+    private OutputContext $context;
```
```suggestion
    /** @readonly */
    private OutputContext $context;
```

keradus commented on src/Console/Output/Progress/DotsOutput.php:
```diff
@@ -125,9 +113,14 @@ public function printLegend(): void
+        $this->output()->write(sprintf("\nLegend: %s\n", implode(', ', $symbols)));
+    }
+
+    private function output(): OutputInterface
```
tbh I don't see value of this method - it's typical proxy method.
yes, i see that one can use this method and have few chars shorter code - but also ppl will be able to use 2 variants in this class to get `OutputInterface` - `$this->output()` or `$this->context->getOutput()`.

btw, I don't get why sometimes
```php
function getOutput(): OutputInterface
```
and sometimes
```php
function output(): OutputInterface
```

Wirone replied:
Yes, it's only shortcut to not repeat `$this->context->getOutput()`. Personally I don't like getters with `get*` prefix, it's redundant for me. Method's name itself should be descriptive enough to tell developer if the intention is to access something (get) or make some action (set).

keradus replied:
pls avoid proxy methods, there should be one way to get given value, otherwise those 2 "for now identically resulted" approaches will diverge over time - and till they will, it will still cause confusion which variant to use.

for whole project, we use `getX` approach.

Wirone replied:
This is private method and is used consistently within whole class (6 times), there's no long form usage. I don't agree about diverging and confusion, and I believe `$this->output()->write()` is a lot more readable and meaningful than `$this->context->getOutput()->write()` 🤷‍♂️. I can change it if you really want, but I would prefer to keep it as-is.

PS. For public APIs I keep `getX()` convention (as you can see in `src/Console/Output/OutputContext.php`), but here I wanted to make code less cluttered.

keradus replied:
let's use `getX` for whole project, as we do right now. including `OutputContext`. Methods should have verbs, properties should not.

keradus commented on src/Console/Output/Progress/ProgressOutputFactory.php:
```diff
@@ -0,0 +1,51 @@
+        }
+
+        if (!self::isBuiltInType($outputType)) {
+            throw new \RuntimeException(
```
`InvalidArgumentException`  ?

Wirone replied:
I'll also change `self::isBuiltInType()` to `$this->isBuiltInType()` because method is not static (it was at some point I believe, but it was refactored several times).

keradus commented on src/Console/Output/Progress/ProgressOutputFactory.php:
```diff
@@ -0,0 +1,51 @@
+    public function create(string $outputType, OutputContext $context): ProgressOutputInterface
+    {
+        if (null === $context->getOutput()) {
+            $outputType = ProgressOutputType::NONE;
```
should we raise exception if `getOutput()===null && outputType !== 'none'` ?

Wirone replied:
It's thrown below, if `$outputType` is not any of supported built-in types. This condition here is only about forcing empty output if there's no CLI output available (honestly I don't remember now why it was done like this, it was 3 months ago).

keradus replied:
I am talking about valid outputTypes. like 'dots'.

especially this test case: https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/pull/6848/files#diff-d9c809453496cbba84934580e854433ee280b478095ad63c62338c51f9c78cb2R52
you silently called back to `none` output type, i wonder about throwing exception of invalid input. would that make sense?

Wirone replied:
Exceptions are for... exceptional situations 😉. If output is no available, we can just fallback to none output, so it makes is non-exceptional IMHO. This condition AFAIR is for some edge cases where `OutputInterface` in the command is not an instance of `ConsoleOutputInterface` and format is other than `txt` - then `$stdErr` is null and then progress output type is overridden to `none`. And that's OK in my opinion. In most scenarios there will be output available.

keradus replied:
i am basically wondering on which level to make that fallback - here, on factory method, or on class facing user input.

Wirone replied:
IMHO this is the best place for it - user's input is resolved in `ConfigurationResolver` but the output is determined in `FixCommand` based on CLI's I/O and resolved config. Here we have central place that can decide which progress output fits best considering the context.

keradus commented on src/Console/Output/Progress/ProgressOutputFactory.php:
```diff
@@ -0,0 +1,51 @@
+            );
+        }
+
+        return ProgressOutputType::NONE === $outputType
```
```suggestion
        /** @TODO: PHP 8.0+ - convert to `match` */
        return ProgressOutputType::NONE === $outputType
```

Wirone replied:
There's no point in adding such comment here because it will be refactored anyway in consequent PR #6842.

keradus replied:
in generla - for reviewing and ultimately aiming for merge of one PR, I do not assume what may or may not be changed in other PR that may or may not get merged 🤷🏻

Wirone replied:
Even without that PR, converting this to `match` does not make much sense with only 2 possible paths - ternary is good enough 😅.

keradus replied:
match is more extendable ;P

keradus reviewed (approved):
a few minor topics could get updated, but I believe we do not need an explicit round of review for them.

Wirone commented:
@keradus thank you for your time, please make decisions on remaining discussions - I've fixed what I agreed with, the others ([1](https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/pull/6848#pullrequestreview-1509121905), [2](https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/pull/6848#pullrequestreview-1509135052) and [3](https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/pull/6848#pullrequestreview-1509141872)) I am reluctant to change, but will do it if you really want it. Just let me know.

keradus commented:
https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/pull/6848#issuecomment-1481938401

Wirone commented:
![](https://media.tenor.com/p66O7j7UylkAAAAC/preacher-hallelujah.gif)
